File: android/app/src/main/kotlin/jp/co/bitz/tictactoe/scenes/MenuScene.kt
package jp.co.bitz.tictactoe.scenes

import android.content.Context
import android.graphics.Color
import jp.co.bitz.spritekit.SKLabelNode
import jp.co.bitz.spritekit.SKScene
import jp.co.bitz.spritekit.SKSceneScaleMode
import jp.co.bitz.spritekit.SKTransition
import jp.co.bitz.spritekit.Vector2
import jp.co.bitz.tictactoe.game.Difficulty
import jp.co.bitz.tictactoe.game.TicTacToeMatch
import kotlin.time.Duration.Companion.seconds

/**
 * The app's entry scene: difficulty picker → coin toss → `GameScene`. See
 * docs/GAME_DESIGN.md's "Presentation" section, `MenuScene`.
 */
public class MenuScene(
    private val context: Context,
    size: Vector2,
) : SKScene(size) {
    private object Layout {
        const val TITLE_Y = 1450f
        const val TITLE_FONT_SIZE = 84f
        val BUTTON_SIZE = Vector2(560f, 140f)
        const val BUTTON_FONT_SIZE = 48f
        const val FIRST_BUTTON_Y = 1050f
        const val BUTTON_SPACING = 200f
    }

    init {
        scaleMode = SKSceneScaleMode.AspectFit
        backgroundColor = Color.BLACK
        buildContent()
    }

    private fun buildContent() {
        val strings = Strings(context)

        val title =
            SKLabelNode(strings.menuTitle).apply {
                fontSize = Layout.TITLE_FONT_SIZE
                fontColor = Color.WHITE
                position = Vector2(size.x / 2f, Layout.TITLE_Y)
            }
        addChild(title)

        val difficulties =
            listOf(
                Difficulty.EASY to strings.difficultyEasy,
                Difficulty.NORMAL to strings.difficultyNormal,
                Difficulty.HARD to strings.difficultyHard,
            )
        difficulties.forEachIndexed { index, (difficulty, label) ->
            val button =
                ButtonNode(
                    size = Layout.BUTTON_SIZE,
                    cornerRadius = 24f,
                    label = label,
                    fontSize = Layout.BUTTON_FONT_SIZE,
                )
            button.position = Vector2(size.x / 2f, Layout.FIRST_BUTTON_Y - index * Layout.BUTTON_SPACING)
            button.onTap = { startGame(difficulty) }
            addChild(button)
        }
    }

    private fun startGame(difficulty: Difficulty) {
        val humanMark = TicTacToeMatch.coinTossMark()
        val nextScene = GameScene(context, size, difficulty, humanMark)
        view?.presentScene(nextScene, SKTransition.crossFade(0.5.seconds))
    }
}

File: ios/TicTacToe/Scenes/MenuScene.swift
import SpriteKit

/// The app's entry scene: difficulty picker → coin toss → `GameScene`. See
/// `docs/GAME_DESIGN.md`'s "Presentation" section, `MenuScene`.
final class MenuScene: SKScene {
    private enum Layout {
        static let titleY: CGFloat = 1450
        static let titleFontSize: CGFloat = 84
        static let buttonSize = CGSize(width: 560, height: 140)
        static let buttonFontSize: CGFloat = 48
        static let firstButtonY: CGFloat = 1050
        static let buttonSpacing: CGFloat = 200
    }

    override init(size: CGSize) {
        super.init(size: size)
        scaleMode = .aspectFit
        backgroundColor = .black
        buildContent()
    }

    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    private func buildContent() {
        let title = SKLabelNode(text: Strings.menuTitle)
        title.fontSize = Layout.titleFontSize
        title.fontColor = .white
        title.position = CGPoint(x: size.width / 2, y: Layout.titleY)
        addChild(title)

        let difficulties: [(Difficulty, String)] = [
            (.easy, Strings.difficultyEasy),
            (.normal, Strings.difficultyNormal),
            (.hard, Strings.difficultyHard),
        ]
        for (index, entry) in difficulties.enumerated() {
            let (difficulty, label) = entry
            let button = ButtonNode(
                size: Layout.buttonSize,
                cornerRadius: 24,
                label: label,
                fontSize: Layout.buttonFontSize
            )
            button.position = CGPoint(x: size.width / 2, y: Layout.firstButtonY - CGFloat(index) * Layout.buttonSpacing)
            button.onTap = { [weak self] in self?.startGame(difficulty: difficulty) }
            addChild(button)
        }
    }

    private func startGame(difficulty: Difficulty) {
        let humanMark = TicTacToeMatch.coinTossMark()
        let nextScene = GameScene(size: size, difficulty: difficulty, humanMark: humanMark)
        view?.presentScene(nextScene, transition: .crossFade(withDuration: 0.5))
    }
}
